src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,528 @@
+		/// <returns>The created transaction context.</returns>
+		protected virtual ITransactionContext CreateAndEnlistMainContext(
+			ISessionImplementor session,
+			System.Transactions.Transaction transaction)
+		{
+			var transactionContext = new SystemTransactionContext(
+				session, transaction, UseConnectionOnSystemTransactionEvents);
+			transactionContext.AmbientTransaction.EnlistVolatile(
+				transactionContext,
+				UseConnectionOnSystemTransactionEvents
+					? EnlistmentOptions.EnlistDuringPrepareRequired
+					: EnlistmentOptions.None);
+			return transactionContext;
+		}
+
+		private void EnlistDependentSession(ISessionImplementor dependentSession, ITransactionContext mainContext)
+		{
+			dependentSession.TransactionContext = CreateDependentContext(dependentSession, mainContext);
+			dependentSession.AfterTransactionBegin(null);
+		}
+
+		/// <summary>
+		/// Create a transaction context for a dependent session.
+		/// </summary>
+		/// <param name="dependentSession">The dependent session.</param>
+		/// <param name="mainContext">The context of the session owning the <see cref="ConnectionManager"/>.</param>
+		/// <returns>A dependent context for the session.</returns>
+		protected virtual ITransactionContext CreateDependentContext(ISessionImplementor dependentSession, ITransactionContext mainContext)
+		{
+			return new DependentContext(mainContext);
+		}
+
+		/// <inheritdoc />
+		public override bool IsInActiveSystemTransaction(ISessionImplementor session)
+			=> session.TransactionContext?.IsInActiveTransaction ?? false;
+
+		/// <inheritdoc />
+		public override void ExecuteWorkInIsolation(ISessionImplementor session, IIsolatedWork work, bool transacted)
+		{
+			using (var tx = new TransactionScope(TransactionScopeOption.Suppress))
+			{
+				base.ExecuteWorkInIsolation(session, work, transacted);
+				tx.Complete();
+			}
+		}
+
+		/// <summary>
+		/// Transaction context for enlisting a session with a system <see cref="System.Transactions.Transaction"/>.
+		/// It is meant for being the concrete class enlisted in the transaction.
+		/// </summary>
+		public class SystemTransactionContext : ITransactionContext, IEnlistmentNotification
+		{
+			/// <summary>
+			/// The transaction in which this context is enlisted.
+			/// </summary>
+			protected internal virtual System.Transactions.Transaction AmbientTransaction { get; }
+			/// <inheritdoc />
+			public bool ShouldCloseSessionOnSystemTransactionCompleted { get; set; }
+			/// <inheritdoc />
+			public bool IsInActiveTransaction { get; protected set; } = true;
+			/// <inheritdoc />
+			public virtual bool CanFlushOnSystemTransactionCompleted => _useConnectionOnSystemTransactionEvents;
+
+			private readonly ISessionImplementor _sessionImplementor;
+			private readonly bool _useConnectionOnSystemTransactionEvents;
+			private volatile SemaphoreSlim _semaphore;
+			private volatile bool _locked;
```

Comment:
> But you already have "_bypassWait"...

Sorry, the mail for this one was lost in spam folder, I am seeing it only now.

I was believing you were suggesting that `_byPassWait` would not be needed with `ManualResetEvent`. So, another misunderstanding. Were you suggesting `ManualResetEvent` for avoiding the `Release(100)`? It would be slightly cleaner, and that was the initial implementation I have done. But in fact we should ever have only two concurrent threads: the one executing after scope disposal, and the one executing transaction completion event. Second phase callback are no more acting on the session.  (And even if we change all that, we should stay at two threads concurrently actually using the session.)  

As said in the comment, the `100` is there only for avoiding freezing unexpected threads launched by buggy user code. So that is just here for a bugged application case anyway, not for a nominal case.

I have switched to `SemaphoreSlim` because they are lighter, they can work without system wide wait handle, and although I have not checked it, they are likely supported in a lower .NetStandard than the full fledged wait handles. We may even asynchronously await it if we deem it preferable for the async PR. (But I would tend to exclude it from async generation instead, since that would be actually useful only on distributed system transaction, which are likely not the main use case of NHibernate.)